Char measurement

as-cii opened the pull request:
Fixes #6055 

This is still a work in progress: code is messy and specs are not passing yet. ~~With this PR we basically wrap every character with a span so that we're able to measure it. This helps to calculate the real character's width on screen: whenever this is not possible, however, the current solution falls back to `@defaultCharWidth`.~~

We have split the `TextEditorPresenter` and the `TextEditorComponent` updates in two phases: pre-measure and post-measure. The former should calculate anything related to lines, whereas the latter  performs updates which need to be based on lines measurement.
- [x] Write new specs to consolidate this behavior
- [x] Make existing specs pass
- [x] Remove references to scoped char width
- [x] Scrolling into an unmeasured line with decorations (or cursors) has issues (since we don't measure such lines on scroll anymore and updating them doesn't require any user action)

as-cii committed:
Wrap every char with a <span>

as-cii committed:
Render paired characters as one <span>

as-cii committed:
Calculate pixel position reading DOM

as-cii committed:
Generate <span> for spaces as well

as-cii committed:
Remove readBeforeUpdateSync as we're not using it now

nathansobo commented on src/display-buffer.coffee:
```diff
@@ -332,6 +322,14 @@ class DisplayBuffer extends Model
     @scopedCharacterWidthsChangeCount++
     @characterWidthsChanged() unless @batchingCharacterMeasurement
 
+  setCharWidthsForLineId: (lineId, charWidths) ->
```
I very much like this approach of just keeping a parallel array.

nathansobo commented on src/display-buffer.coffee:
```diff
@@ -332,6 +322,14 @@ class DisplayBuffer extends Model
+    @charWidthsForLineId[lineId] = charWidths
+    @characterWidthsChanged() unless @batchingCharacterMeasurement
+
+  getCharWidthForLine: (line, index) ->
```
I wonder if this should just be `getCharacterWidth` and take a `Point`. It could then consult its lines internally.

as-cii replied:
:+1:

I'll do that once we agree on pushing this solution forward. What do you think?

nathansobo commented on src/lines-component.coffee:
```diff
@@ -270,8 +270,19 @@ class LinesComponent
+      characters = lineNode.querySelectorAll("span.char")
+      charWidths = []
+
+      for character in characters
```
Take a look at the previous code for an approach to measurement that doesn't depend on spans. You can build an iterator over the text nodes then set the start and end index of a DOM selection for a given pair of offsets in a text node to get a bounding rectangle from that.

nathansobo commented:
I'm wondering if rather than storing the widths of each character at each position if it wouldn't be easier to just store a running total. So if every character were 10px wide on a line with 4 characters, it would be `[10, 20, 30, 40]`. Basically, rather than an array of widths it is now an array of the horizontal position of the right side of that character in absolute terms.

nathansobo commented:
:rocket: Looking forward to collaborating on this problem with you.

as-cii committed:
Measure chars using DOM ranges

as-cii committed:
Set char widths by row

as-cii committed:
Calculate char widths for visible rows only

as-cii committed:
Delete any reference to scoped char width

as-cii committed:
Revert 'Calculate char widths for visible rows'

as-cii committed:
Use @newState while iterating over lines

as-cii committed:
:racehorse: Fix typo

as-cii committed:
Merge branch 'master' into as-char-measurement

as-cii committed:
Use char index instead of buffer column

as-cii committed:
WIP: Experiment with (pre/post) updates

This eliminates visual glitches altogether. Code-wise is very messy, but
apparently this may be a good way to measure characters :+1:

as-cii committed:
:art: Split pre/post measure appropriately

as-cii committed:
Always calculate decorations

Because we need to add css classes to lines when they're rendered

as-cii committed:
Update content state in preMeasure

as-cii committed:
Avoid reading after updateSync

as-cii committed:
:fire: Remove conditional and use functions

as-cii commented:
> I'm wondering if rather than storing the widths of each character at each position if it wouldn't be easier to just store a running total. So if every character were 10px wide on a line with 4 characters, it would be [10, 20, 30, 40]. Basically, rather than an array of widths it is now an array of the horizontal position of the right side of that character in absolute terms.

I was just thinking about this and it's definitely something we could do. I wonder, however, if we could make it even faster :open_mouth: What if, while measuring lines, we stored:

* A matrix of rows and columns, with the absolute left position of each column
* An interval skip list (or a similar data structure) of absolute left positions (this is needed for clipping when in the middle)

If we manage to do this, then `screenPositionForPixelPosition` and `pixelPositionForScreenPosition` would basically avoid tokens iteration altogether, making finally use of the existing computation performed by `measureCharactersInVisibleLines`. What do you think @nathansobo? This would be huge in terms of speed! :racehorse: 

By the way, specs are :green_heart: now (exception made for those referring to scoped char widths) and it would be great if you could have a second look at this. Thank you :bow:

In the meantime I'll try to put a bit of thought into a reflow-less solution, even though I feel like this PR performs quite well.

as-cii commented:
@nathansobo: tomorrow I'll write a bunch of specs for this and collect some useful metrics in order to make a more informed decision :rocket:

as-cii committed:
Use only batched updateCursorsState

This way we prevent cursors from calculating a wrong position because of
out-of-sync lines measurement: we need to perform such updates inside
`getPostMeasureState`, where we can guarantee correct char widths.

as-cii committed:
:racehorse: Measure only when editor is visible and not scrolling

as-cii committed:
:racehorse: Handle characterWidthsChanged as before

as-cii committed:
:art:

as-cii committed:
:racehorse: Measure only changed lines

as-cii committed:
Invalidate char widths when changing content

as-cii committed:
:green_heart: Fix TextEditorPresenter specs

as-cii committed:
:green_heart: Fix DisplayBuffer specs

nathansobo commented:
> An interval skip list (or a similar data structure) of absolute left positions (this is needed for clipping when in the middle)

This seems like it might be more than is needed. If you just store a running total in an array, it would be pretty easy to binary search it when converting from pixels to columns.

I'd love to see some profiling / timeline results for this change at some point. I know we need to fix this issue but I'm worried about regressing performance. Thanks so much!

nathansobo commented on src/display-buffer.coffee:
```diff
@@ -37,7 +37,7 @@ class DisplayBuffer extends Model
   verticalScrollMargin: 2
   horizontalScrollMargin: 6
-  scopedCharacterWidthsChangeCount: 0
+  characterWidthsDidChange: false
```
Maybe `characterWidthsChanged`? We have been using `will` and `did` in the names of hooks.

nathansobo commented on src/lines-component.coffee:
```diff
@@ -302,29 +322,24 @@ class LinesComponent
+        i = charIndex - textNodeIndex
+        rangeForMeasurement.setStart(textNode, i)
+        rangeForMeasurement.setEnd(textNode, i + charLength)
+        charWidth = rangeForMeasurement.getBoundingClientRect().width
```
Glad you're using this approach instead of spans. You could leave the start of the measurement range at index 0 of the first text node if you want to store a running total rather than individual widths.

as-cii committed:
Measure while scrolling

as-cii commented:
I have performed some benchmarks and captured some screenshots so that we can consider them together.

All the benchmarks were performed on a 1920x650 px window.

## Move 300 cursors simultaneously

**master**
![master](https://cloud.githubusercontent.com/assets/482957/6853261/6f7ec02a-d3ea-11e4-9ed7-6ff17d898832.png)

**as-char-measurement**
![dev](https://cloud.githubusercontent.com/assets/482957/6853269/7e744064-d3ea-11e4-992a-0c3d3fae3639.png)

**Conclusion**: we have significantly reduced the amount of time spent in updating cursors. This has nothing to do with our char measurement strategy, however, because it's mainly due to 8a153f16596ba2c334a6c5e08dc2d942e60a4dbc :tada: 

## Copy and paste a paragraph till we reach ~300 lines

**master**
![master](https://cloud.githubusercontent.com/assets/482957/6853314/d6f0f8d6-d3ea-11e4-91eb-66c950b6d928.png)

**as-char-measurement**
![dev](https://cloud.githubusercontent.com/assets/482957/6853317/ddd2136a-d3ea-11e4-9bec-724efe6c19a7.png)

**Conclusion**: I cannot seem to find any significant difference between the two. Fps are similar too and therefore I think we haven't hit a performance regression here :tada: 

## Scroll down line-by-line on ~200 lines

**master**
![master](https://cloud.githubusercontent.com/assets/482957/6853403/4a7e79a4-d3eb-11e4-97db-5dde823cca68.png)

**as-char-measurement**
![dev](https://cloud.githubusercontent.com/assets/482957/6853410/52ed020e-d3eb-11e4-8a62-4e03d014e9ef.png)

**Conclusion**: I'd say that we have a performance degradation here, but it's not really significant. Earlier on I had introduced a performance optimization to avoid calculating lines widths while scrolling or for previously calculated lines. The main drawback of this approach is that we cannot apply it to cursor scrolling. One approach may be to simply skip already measured lines: however we have to keep in mind that sub-pixel rendering may occur if, for example, a line has changed position, and therefore we need a always fresh value.

## Scroll down via mouse on ~3000 lines

**master**
![master](https://cloud.githubusercontent.com/assets/482957/6853588/5e4cdefc-d3ec-11e4-8357-c81057f70538.png)

**as-char-measurement**
![dev](https://cloud.githubusercontent.com/assets/482957/6853592/6c12df0a-d3ec-11e4-85d7-0b43b5bcbcbe.png)

**Conclusion**: the same considerations of the above scenario apply here.

----------------------------

All in all, I'd say that the editor feels quite responsive. It's definitely a pity to not be able to measure text before actually adding it to the DOM, but I am not sure we have many alternatives here.

@nathansobo: I recall having a conversation with you about the usefulness of `<canvas>`. I know, it would be a quite drastic change, but it would allow us to [measure text](http://www.w3schools.com/tags/canvas_measuretext.asp) before actually rendering it on screen. That API does not seem to have sub-pixel rendering support: however, it's something we could workaround ourselves [by turning floating points into integers](http://www.html5rocks.com/en/tutorials/canvas/performance/) (see **Avoid floating point coordinates**), so that we can be sure that no mismatch occurs between our measured text and the rendered one. Something to keep in mind for the future?

I'd :heart: to have your feedback on this. Thank you :bow:

nathansobo commented:
Looking good. What does typing a character look like?

nathansobo commented:
Scrolling down via mouse looks to have slowed down quite a lot. I wonder if the presenter could have some state that maintains what positions we'll actually need to measure and only measure those locations. Something like `screenRangesToMeasure` or something.

as-cii commented:
> I wonder if the presenter could have some state that maintains what positions we'll actually need to measure and only measure those locations. Something like `screenRangesToMeasure` or something.

That's an optimization we could make and I had that in mind too, but quoting what I've written in my previous comment:

> One way to gain some speed here may be to simply skip already measured lines: however, we need to keep in mind that **sub-pixel rendering may occur** if, for example, a line has changed position, and therefore we need an always fresh char width value.

So, basically, even though we have already measured characters' width for a certain line, when scrolling up or down we have no guarantee that such widths are correct because sub-pixel rendering *may* have happened. What do you think?

as-cii committed:
Merge branch 'master' into as-char-measurement

as-cii committed:
Handle changed lines in TextEditorPresenter

as-cii committed:
Avoid measuring while scrolling

as-cii committed:
:white_check_mark: Write specs for char measurement on scrolling

as-cii committed:
Move ::getPreMeasureState out in its own describe

as-cii committed:
:art: Rename to handleCharacterWidthsChanged

as-cii committed:
:white_check_mark: Account for larger characters

as-cii commented:
![screen shot 2015-03-27 at 09 48 35](https://cloud.githubusercontent.com/assets/482957/6866570/e382b06e-d479-11e4-817e-0c1abaace168.png)

While scrolling via mouse we're now operating as fast as **master** :chart_with_upwards_trend: :chart_with_upwards_trend: :chart_with_upwards_trend:

Basically, we now simply avoid to mark lines as changed while scrolling: we just keep in mind that we'll need to change them once scrolling ends in `TextEditorPresenter`. This solves also the sub-pixel rendering issue I was mentioning earlier on, because right before any user action (after scrolling has ended) we will compute the char widths.

Specs are :green_heart: and I wrote some new ones for this "changed lines performance improvement". I think, however, that we need to update all the existing specs in `text-editor-presenter-spec.coffee` to use `getPreMeasureState` and `getPostMeasureState`, otherwise they won't honor the *real* contract between `TextEditorPresenter` and `TextEditorComponent`.

@nathansobo: this afternoon I will have limited availability, therefore please feel free to jump on this if you want. I feel like we're quite close to :shipit: and I am looking forward to hearing what you think.

Moreover, I'd like to try this out with sub-pixel rendering enabled: any ideas how we could do that (since it will involve patching again `libchromiumcontent`)?

Thank you!

as-cii committed:
Merge branch 'master' into as-char-measurement

as-cii committed:
Dummy canvas

as-cii committed:
Store contexts by scopes

as-cii committed:
Clear @contextsByScopeIdentifier

as-cii committed:
:racehorse: Position <canvas> in iframe

as-cii committed:
Update only needed state in the postMeasure phase

as-cii committed:
:racehorse: Assign IDs to tokens

as-cii committed:
Store total left

as-cii commented:
> Now you can actually just iterate over leaf spans with unique scopes instead of text nodes since you only need computed styles. That could probably save some work.

I ended up assigning a unique id to each token. I am not super happy with how I access it, given that I use `document.querySelector` (with the `/deep/` marker). All in all, however, it performs quite nicely and it's faster compared to the iterator-based version. I am open to suggestions on how to make this even better.

Our most performance critical area is scrolling: one thing to keep in mind is that now we do an additional operation (i.e. measuring lines) while rendering lines, so it'll be difficult to reach the same exact level of smoothness we had before these changes.

*(Benchmarks performed on `text-editor-component-spec.coffee`)*

**master**
![screen shot 2015-04-01 at 18 44 05](https://cloud.githubusercontent.com/assets/482957/6946256/47021a98-d89f-11e4-97f6-0f501a34480b.png)

**as-char-measurement**
![screen shot 2015-04-01 at 18 37 52](https://cloud.githubusercontent.com/assets/482957/6946257/47310c40-d89f-11e4-9fe4-b144d3eacd24.png)

There's a significant performance boost compared to the graphs I showed earlier on. Still, however, we're not always operating at 60fps but I do think we've got several things we may do better. I'll write below a series of facts and, sometimes, possible solutions to explore.

* Things get garbage collected during animation frames. It is the cause of some of those spikes you can see in the graph, but I still need to perform benchmarks to see how memory is actually allocated and how we can make a better use of it.
* On average, measuring lines usually takes less than 6ms. I am not sure how much we can reduce this: in other words it means that we may have only ~10ms to perform everything else.
* Updating scroll top causes both a reflow and a layout (~7ms). What about ditching the default scrollbar for a virtual one (which would not trigger any expensive operation)? Positioning elements will become slightly more complicated, but we could save something by doing that. 
* The remaining time is spent updating line nodes (~11ms). I conducted a quick experiment with `DocumentFragment`: apparently, however, creating nodes by hand is slower than appending strings as we do now. I must admit I didn't dig as much as I'd have liked on this, but it would be really awesome to find a cheaper alternative. An idea could be to have a pool of objects, instead of  always creating and removing new lines (this could even solve some GC slowdowns?), but I haven't any metrics at hand to confirm it would work.

@nathansobo: I know there's quite a lot of stuff in this comment but I wanted to keep you posted about this. I am going to spend some more time on it, especially trying out some of the optimizations I outlined above. I am sorry if this is taking longer than expected: I am truly looking forward to see this :ship:ped as soon as possible, and I feel like we're halfway there :fire:. 

Any suggestion or comment is always welcome and appreciated :bow: :bow: 

Thank you.

nathansobo commented:
No worries about timing. I'm keeping an eye on this and want to support you, but this is still elective surgery for at least a few more Chrome releases.

as-cii committed:
:fire: Remove leftover specs

as-cii committed:
Merge branch 'master' into as-char-measurement

# Conflicts:
#	spec/text-editor-presenter-spec.coffee
#	src/lines-component.coffee
#	src/text-editor-component.coffee
#	src/text-editor-presenter.coffee

as-cii committed:
:bug: Avoid measuring only when not scrolling

nathansobo commented:
@as-cii What's the status on this? Do you think this is ready to focus on merging? As long as we don't lose performance I think it would be better to fix this sooner rather than later.

as-cii committed:
Merge branch 'master' into as-char-measurement

as-cii committed:
:racehorse: Build scopesIdentifier in Token

as-cii committed:
Use points instead of plain arrays

as-cii committed:
Early return on TextEditorComponent#pollDOM during scrolling

as-cii committed:
:racehorse: Avoid measuring already measured text

as-cii committed:
Disable subpixel-font-scaling

We should enable it once we're sure this new measurement works correctly.

as-cii committed:
:racehorse: Allow compiler optimizations while iterating hashes

as-cii committed:
Measure only when lines stay on screen for > 2 frames

Broadly speaking, measuring characters is a quite cheap task (especially since
we introduced <canvas> and its `measureText` API). However, when scrolling we
repeat this action several times and it's likely for a frame to miss the 16ms
goal: in other words, this means that we cannot afford to always measure newly
built lines.

For this reason, this commit introduces a new algorithm which chooses when
characters should be measured. While scrolling we'll now check if the line is on
screen for at least 2 frames (we can even make this configurable): this will
avoid "measurement thrashing" and, hypothetically, the user won't notice that
we're not measuring a certain line. Benchmarks are quite promising and,
apparently, performance don't degrade at all after these changes.

An alternative implementation could have been to decrease the `didStopScrolling`
timeout and measure right after scrolling stopped: however, that would yield to
a less accurate timing (which is not in sync with `requestAnimationFrame`) and
increases the likelihood of noticing a misplaced graphic element (e.g.
decoration).

Please note that this is still not polished and there are some edge cases which
may need to be addressed. On the other hand it demonstrates what I had in mind
so that we're able to validate this proposal based on real code.

/cc: @nathansobo

as-cii committed:
Merge branch 'master' into as-char-measurement

as-cii committed:
:green_heart: Fix DisplayBuffer spec failures

as-cii committed:
Remove already measured text optimization

as-cii committed:
:green_heart: Fix TextEditorComponent spec failures

There are still three failing tests which seem related one another.

as-cii committed:
:green_heart: Fix TextEditorPresenter spec failures

as-cii committed:
:art: Rename to setCharLeftPositionForPoint

as-cii committed:
:art: Refactor text measurement

as-cii committed:
:art:

as-cii committed:
Remove scope nesting

as-cii committed:
Query token inside a specific line

as-cii committed:
Fix Token HTML specs

as-cii commented:
@nathansobo: as discussed yesterday, I am still working to make this mergeable and this will hopefully happen by the end of this week :fire: 

Right now there are still a few things I need to address. Specifically:

1. [ ] :green_heart: Look into two tests that are currently failing for no apparent good reason (it should be quite straightforward, though);
2. [ ] :racehorse: Avoid computing scroll width in `DisplayBuffer` every single time `setCharLeftForPoint` is called;
3. [ ] :bug: Scrolling from model (i.e. `editor.setScrollTop`) should trigger a character measurement on newly built lines;
4. [ ] Resolve conflicts with `master`

In the meantime, however, it would be great if you could start having a look at this and leave some comments wherever you feel we should change something. Code-wise this is not perfect yet and, since it has been up for a while and I experimented a lot, there may be a bunch of things we want to refine and do better.

Thanks! :bow:

as-cii commented on src/display-buffer.coffee:
```diff
@@ -293,36 +293,39 @@ class DisplayBuffer extends Model
-    @getScopedCharWidths(scopeNames)[char] = width
-    @scopedCharacterWidthsChangeCount++
-    @characterWidthsChanged() unless @batchingCharacterMeasurement
+  setCharLeftPositionForPoint: (row, column, charWidth) ->
```
I am not super happy with the name of this method. I couldn't come up with something clever and everything always felt like a mouthful. Any ideas how could we improve it? :thought_balloon:

nathansobo replied:
Only a slight improvement: `setLeftPixelPositionForPoint`. I think emphasizing *pixels* is a good idea.

as-cii committed:
Merge branch 'master' into as-char-measurement

# Conflicts:
#	src/lines-component.coffee
#	src/text-editor-presenter.coffee

aceLren commented:
Will this address #3821?  Ligatures broke a few versions back and it would be great to have it for fonts like Hasklig and Fire Code.

nathansobo commented:
Yeah, it should fix ligatures. We need to try something more efficient than the current state of this PR because it has a bit too much overhead.

nathansobo commented:
Wanted to provide a status update on this because implementing correct character measurement is becoming a bottleneck for other efforts, such as CJK-aware softwrap. @as-cii feel free to chime in and fill in any details I may be missing.

@as-cii and I had a screen sharing session about this PR a few weeks back, and concluded that the measurement overheads for the current state of this PR are probably too high. Our overall goal is to be able to compute the horizontal pixel position of a column at any time without reading from the document (which causes reflows).

* The existing solution on master caches the width of each character in each scope whenever the DOM is updated, allowing us to compute the horizontal location of a column by summing up characters and reading out of this cache. Unfortunately, this approach is broken with ligatures and doesn't work when subpixel font scaling is enabled.

* The solution on this PR is to use the canvas API to measure text. The advantage to measuring via canvas is that we don't interact with the document at all, but unfortunately, the canvas text measurement API seems to be too slow for our purposes, requiring every prefix of the line to be measured in a separate call.

* The idea @as-cii and I explored on our call is to instead perform measurement in a separate document, via an iframe. We would need to *sample* styling information for every unique scope in the document via dummy DOM nodes, similar to @abe33's approach for the [minimap](https://github.com/abe33/atom-minimap). Of course all we're interested in is font styling, since that's the only thing that affects measurement. Then, to measure the position of all columns on a line, we would construct a simpler version of that line in a different document with the same styling applied inline, then use the DOM range API to iterate over the characters and measure. The advantage of this is that we only need to perform layout once.

Will this new approach work? Not sure. But it's the next thing we'd like to try. Other ideas are welcome, but remember that it can't rely on reading from the document.

superlou commented:
@nathansobo: Will this approach allow positioning on lines that have margins or padding, or allow some other means of changing softwrap boundaries throughout the text editor?

nathansobo commented:
@superlou I'm not sure how much CSS applied to lines it's going to be realistic to support. Any supported styling will need to be sampled for measurement in the other document, and covering every possible thing people may want to style positioning wise could end up being pretty complicated. So we can try, but we may need to stick to CSS for styling fonts and colors and add other editor-specific mechanisms for controlling soft wrap behavior etc. We'll see how it goes.

larsenwork commented:
Not sure if it's related to this but I'm having issues with opentype `calt ` stopped working

nathansobo commented:
@larsenwork Nothing has really changed in this area for a while, so I'd say it's unlikely. You can still open another issue however.

alexandernst commented:
Is this still a thing?

nathansobo commented:
@alexandernst Yep. It's still happening. @as-cii has been churning through a zillion alternatives looking for the best approach. It's looking increasingly like just eating a small forced reflow might be the best approach. We're getting there.
